token_account_storage/src/inmemory_token_storage.rs

```diff
@@ -495,6 +580,54 @@ impl TokenProgramAccountsStorage {
                                         acc.owner == owner
                                     }
                                 })
+                            {
+                                if let Some(account_data) = token_account_to_solana_account(
+                                    token_account,
+                                    finalized_slot,
+                                    0,
+                                    &self.mints_by_index,
+                                ) {
+                                    let mint_account_ref =
+                                        self.mints_by_index.get(&token_account.mint).expect("Mint must exist when a token account for a mint exists");
+                                    let mint_account = mint_account_ref.value();
+                                    accounts.push(TokenProgramAccountData::TokenAccount(
+                                        TokenProgramTokenAccountData {
+                                            account_data,
+                                            mint_pubkey: mint_account.pubkey,
+                                        },
+                                    ));
+                                    if mints.contains_key(&mint_account.pubkey) {
+                                        mints.insert(mint_account.pubkey, mint_account.clone());
+                                    }
+                                }
+                            }
+
+                            Ok((accounts, mints))
+                        }
+                        None => Ok((vec![], HashMap::with_capacity(0))),
+                    }
+                } else if let Some(mint) = mint_filter {
+                    // filter token accounts by mint
+                    match self
+                        .mints_index_by_pubkey
```

Comment:
this index indirection trick should go into a helper method (or wrapping type);
the two Optionals must be treated differently:
`None` from `mints_index_by_pubkey` means **not found** while `None` from `accounts_index_by_mint` points to an inconsistency